### PR TITLE
Adjustable pagination

### PR DESCRIPTION
**Pull request by martijnbleeker**
Add adjustable page sizes for pagination.

If `page_size_options` is set in a controller, the index view will get page size options next to the page selection.

Example for `page_size_options`:
``` ruby
page_size_options [10, 30, 50, 100, 200]
```

- [x] Fix bug where filters are cleared on page_size change
- [x] Fix bug where page_size is cleared on filter change
- [x] Fix bug where page_size selection is visible on empty list

### CODE REVIEW

File: core/lib/brightcontent/base_controller_ext/pagination.rb [hunk 1]
```diff
@@ -5,21 +5,61 @@ module BaseControllerExt
...
           @per_page_count = number
         end
+
+        def page_size_options(sizes)
+          return unless sizes.is_a? Array
+          @per_page_sizes =
+            sizes.select { |size| size.is_a?(Integer) && size > 0 }
+        end
+
+        def per_page_sizes
```
Comment (jvdp) [inline core/lib/brightcontent/base_controller_ext/pagination.rb]: There's a slight inconsistency in naming here: `per_page_count` vs `per_page_sizes` vs `page_size_options`. (This was there previously as well with `per_page_count` the instance variable and `per_page` the class method.)

Perhaps we should use `page_size` and `page_size_options` throughout?

File: core/app/views/brightcontent/base/index.html.erb [hunk 2]
```diff
@@ -19,5 +19,7 @@
 </div>
 
 <div class="panel-footer panel-footer-pagination">
-  <%= will_paginate collection, renderer: BootstrapPagination::Rails %>
+  <%= render partial: "page_sizes",
+    locals: { sizes: per_page_sizes, current_size: items_per_page } %>
```
Comment (jvdp) [inline core/app/views/brightcontent/base/index.html.erb]: Why do you pass these in as locals?

File: core/app/views/brightcontent/base/_page_sizes.html.erb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+<% if sizes.try(:any?) %>
```
Comment (jvdp) [inline core/app/views/brightcontent/base/_page_sizes.html.erb]: `per_page_sizes` defaults to an array so the nil check isn't necessary. Also, if there is only one element these options don't need to be shown either.

File: core/lib/brightcontent/base_controller_ext/pagination.rb [hunk 4]
```diff
@@ -5,21 +5,61 @@ module BaseControllerExt
...
         end
 
         def per_page(number)
           @per_page_count = number
         end
+
+        def page_size_options(sizes)
+          return unless sizes.is_a? Array
+          @per_page_sizes =
+            sizes.select { |size| size.is_a?(Integer) && size > 0 }
```
Comment (jvdp) [inline core/lib/brightcontent/base_controller_ext/pagination.rb]: How realistic is it that a developer will enter bad values here?

File: core/lib/brightcontent/base_controller_ext/pagination.rb [hunk 5]
```diff
@@ -5,21 +5,61 @@ module BaseControllerExt
...
+        helper_method :current_page
+        helper_method :per_page_sizes
+        helper_method :corrected_page
+      end
+
       module ClassMethods
         def per_page_count
-          @per_page_count ||= 30
+          @per_page_count =
+            @per_page_sizes.try(:any?) ? @per_page_sizes[0] : (@per_page_count || 30)
```
Comment (jvdp) [inline core/lib/brightcontent/base_controller_ext/pagination.rb]: I think `@page_sizes` can replace the `@per_page_count` variable; an array with one element will have the same effect and can remove the need for nil and empty checks in other places.

**Comment by martijnbleeker** [comment]
Hi @jvdp, thanks for your feedback.
I've processed it and committed the changes.
Also, we've encountered some minor bugs in the meantime,
those have been fixed as well. (See checkboxes in the initial PR comment)

File: core/app/views/brightcontent/base/index.html.erb (hunk 2 above)
Comment (martijnbleeker) [inline core/app/views/brightcontent/base/index.html.erb]: I took another look and I think it's not really necessary here.
Let's call it a personal preference.

**Comment by jvdp** [comment]
Thanks, looks good!

**Pull request merged**